staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -437,35 +448,86 @@ func (p *patcher) applyPatch(_ context.Context, _, currentObject runtime.Object)
+		return nil, err
+	}
+	if _, err := p.applyAdmission(ctx, objToCreate, nil); err != nil {
+		return nil, err
+	}
+	attributes := p.admissionAttributes(ctx, objToCreate, nil)
+	createValidation := rest.AdmissionToValidateObjectFunc(p.admissionCheck, attributes)
+	return p.restApplier.Create(ctx, objToCreate, createValidation, false)
+}
+
+func (p *patchCreater) admissionType() admission.Operation {
+	return admission.Create
+}
+
+func (p *patcher) admissionAttributes(ctx context.Context, updatedObject runtime.Object, currentObject runtime.Object) admission.Attributes {
+	admissionOperation := p.operation.admissionType()
+	userInfo, _ := request.UserFrom(ctx)
+	return admission.NewAttributesRecord(updatedObject, currentObject, p.kind, p.namespace, p.name, p.resource, p.subresource, admissionOperation, userInfo)
+}
+
 // applyAdmission is called every time GuaranteedUpdate asks for the updated object,
 // and is given the currently persisted object and the patched object as input.
 func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Object, currentObject runtime.Object) (runtime.Object, error) {
 	p.trace.Step("About to check admission control")
-	return patchedObject, p.admissionCheck(patchedObject, currentObject)
+	mutatingAdmission, ok := p.admissionCheck.(admission.MutationInterface)
+	if ok && mutatingAdmission.Handles(p.operation.admissionType()) {
+		attributes := p.admissionAttributes(ctx, patchedObject, currentObject)
+		return patchedObject, mutatingAdmission.Admit(attributes)
+	}
+	return patchedObject, nil
 }
 
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, error) {
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
-		p.mechanism = &jsonPatcher{patcher: p}
+		mechanism := &jsonPatcher{patcher: p}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, err
 		}
-		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		mechanism := &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, model: scope.OpenAPISchema}
+		_, err := p.restPatcher.Get(ctx, p.name, &metav1.GetOptions{})
+		if errors.IsNotFound(err) && p.restApplier != nil {
+			// Only attempt to create if the object doesn't exist and the storage implements Applier
+			// since Patcher is not sufficient to allow the Create operation
+			p.operation = &patchCreater{patcher: p}
```

Comment:
A few questions about transforming patches into updates:
* how does this interact with strategy AllowCreateOnUpdate?
* is the patch verified to have the same name and not make use of generateName?
* does this mean anyone with patch permissions can now create objects?

Reply:
if we determine we're in a create path, we have the request attributes and information required to check if they have create permissions (and as an extra bonus, can include the resource name in the check). We only need to do the check once (not on every internal retry in conflict cases), and that lets us continue distinguishing between creating and updating objects (in most cases both will be granted together, but not all)

Reply:
> how does this interact with strategy AllowCreateOnUpdate?

my initial thought is that we should drop that from the strategy interface, and do the same secondary create authorization check in create-on-update cases.

I think we want to be able to count on server side apply being usable for idempotent crupdate across objects. As long as authorization, admission, validation and the rest storage all see creates as creates, I think our bases are covered for apply/patch

one question would be about rest storage that implemented custom/different logic in their create and update methods (service registry, I'm looking at you), if update was called on a non-existent object. We'd need to sweep/think through the implications there. 

We'd also need to make sure the namespace lifecycle admission plugin correctly prevented create-via-patch and create-via-update in terminating namespaces

Reply:
> my initial thought is that we should drop that from the strategy interface, and do the same secondary create authorization check in create-on-update cases.
> 
> I think we want to be able to count on server side apply being usable for idempotent crupdate across objects. As long as authorization, admission, validation and the rest storage all see creates as creates, I think our bases are covered for apply/patch

It would probably be good to make sure there's agreement on this issue (sig-arch/sig-apimachinery, etc) since it would remove power of individual handlers to opt out of crupdate behavior, for the benefit of consistent apply. 

Ok, I'm done talking to myself now.

Reply:
> how does this interact with strategy AllowCreateOnUpdate?
I think I agree with getting rid of AllowCreateOnUpdate. But does removing it needs to be done in this PR? Create on Update was already possible with patch requests if AllowCreateOnUpdate is true for the strategy. Let me know if you want me to add a TODO or create a follow up issue for it, or if you think it needs to be addressed as part of this.

> is the patch verified to have the same name and not make use of generateName?
I'm not sure I understand what apply is supposed to do with generated name, since patch has to to be called with a name as part of the path. Would you expect it to just ignore the name provided in the path?
I'll look into what kubectl apply does when you tell it to generate a name.

> does this mean anyone with patch permissions can now create objects?
I'm pretty sure this won't cause that to be allowed, since in the create path, the validation is done with admission.Create as the operation. but I'm making sure now, and I'll definitely add a test for it because that is really important.

Reply:
> I think I agree with getting rid of AllowCreateOnUpdate. But does removing it needs to be done in this PR? Create on Update was already possible with patch requests if AllowCreateOnUpdate is true for the strategy. Let me know if you want me to add a TODO or create a follow up issue for it, or if you think it needs to be addressed as part of this.

in this PR, I think we should either honor or get rid of strategy.AllowCreateOnUpdate(). Trying to create via Update (removing the "must exist" check for the apply mechanism inside `applyPatch`) will honor it automatically.

> I'm not sure I understand what apply is supposed to do with generated name, since patch has to to be called with a name as part of the path. Would you expect it to just ignore the name provided in the path?
> I'll look into what kubectl apply does when you tell it to generate a name.

I'd expect name to be set and match the request attributes. Just making sure someone can't patch "/pods/foo" with an apply patch of `{"metadata":{"name":"","generateName":"bar"}}`

> I'm pretty sure this won't cause that to be allowed, since in the create path, the validation is done with admission.Create as the operation. but I'm making sure now, and I'll definitely add a test for it because that is really important.

by the time you get to this handler, you're already past all authorization checks